doc/policy/packages.md

```diff
@@ -117,3 +118,36 @@ backward-compatible for users and applications that rely on p2p transaction rela
 package validation should not prevent the acceptance of a transaction that would otherwise be
 policy-valid on its own. By always accepting a transaction that passes individual validation before
 trying package validation, we prevent any unintentional restriction of policy.
+
+### Package Replace by Fee
+
+Child-with-unconfirmed-parents packages may replace mempool transactions under the following conditions:
+
+1. The child transaction does not conflict with any transactions in the mempool.
+
+*Rationale*: Reduce complexity. Package RBF is intended for scenarios in which the user cannot
+change the parent transaction (e.g. they were presigned with an untrusted party), but has control
+over the child. For example, Package RBF can be used to replace the commitment transaction of an LN
+counterparty by attaching a high-fee child to an anchor output. If the user later wants to update
+the child (e.g. a further fee-bump), that can be done using a single transaction.
+
+2. The directly conflicting transactions all signal replaceability explicitly.
+
+3. All transactions in the package only include an unconfirmed input if that input was included in
+   one of the directly conflicting transactions or is from another transaction in the package. This
+   rule applies to the transactions that do not directly conflict with any mempool transactions.
+
+4. The package fee pays an absolute fee of at least the sum paid by the original transactions.
+
+5. The additional fees (difference between absolute fee paid by the package after deduplication and the
+   sum paid by the original transactions) pays for the package's bandwidth at or above the rate set
+   by the node's incremental relay feerate. For example, if the incremental relay feerate is 1
+   satoshi/vB and the package after deduplication contains 500 virtual bytes total, then the package
+   fees after deduplication is at least 500 satoshis higher than the sum of the original transactions.
+
+*Rationale*: These fee-related rules are identical to the replacement [policy for individual
```

Comment:
Since we are introducing a new form of RBF that doesn't currently exist, now might be the best time to at least improve the BIP 125 rules to be incentive compatible in all situations (though this comes at the cost of making replacements potentially more expensive).

In particular what I have in mind is that we require the ancestor feerate of the child transaction to be higher than the actual feerate of all transactions that would be evicted. This is conservative but I think would ensure that we don't have any situations where there are transactions being replaced that might appear in the next block, while the new transactions do not.

Would this be too conservative for the use cases that we have in mind?

Reply:
Can you elaborate with an example? It's possible I don't have the same mental mapping to terms and do not understand the scenario.

Reply:
Sure -- consider the case where you have a transaction A(100vb, 1sat/vb) and child B (100vb, 10sat/vb) in the mempool.  Total fees = 1100 sat.

A package comes in with parent A'(100vb, 1sat/vb) and child B'(1000vb, 2.5sat/vb).  Total package fees = 2600 sat.

Under the rules described here, package (A', B') could replace (A, B) in the mempool, because total fees are paid for and there's enough left over fee to pay the incremental relay fee of the new package (1500 sat/1100 vb). Also the package feerate is higher than that of the direct conflicts[1].

The deficiency in the logic is that we're not looking at the feerates of the transactions that would be evicted -- even though their absolute fee is being paid, their feerate could be even higher than the transactions that would replace them.  Calculating the effective feerate that such a transaction would have in our mining code is tricky (because a sibling could pay for an ancestor, it's not quite right to just consider ancestor feerate), but using it's feerate is a conservative check here -- any descendant that is paying for its confirmation would be caught in this logic, and it's (higher) feerate considered if necessary.

For evaluating the mining suitability of the new transactions coming in, I think we can just look at the ancestor feerate of the package, which is a lower bound on the attractiveness of the new transactions.  (I think this would also allow us to relax the rule against new unconfirmed inputs.)

[1] In looking at this example, I noticed that BIP 125 as drafted seems to misstate an important check we perform in our current RBF policy, namely that we compare the feerate of a new transaction coming in to the feerate of the existing transactions that it directly conflicts with.  (Without this check, a small high fee transaction could be replaced by a large low fee transaction!). From my first glance at the code in this PR, it looks like the code itself does a similar check for packages (comparing the package feerate to the feerate of the direct conflicts) so that is good at least, but the description of the logic in the documentation should point out that there is some feerate check happening as well.

Reply:
It's ancient lore by now, but I'm not sure it's a "mis-statement" or just incentive incompatible stuff no one implements(becase feerate check is cheap). I agree with your reading of current texts either way.

Reply:
> require the ancestor feerate of the child transaction to be higher than the actual feerate of all transactions that would be evicted. This is conservative but I think would ensure that we don't have any situations where there are transactions being replaced that might appear in the next block, while the new transactions do not.

I think this is a sensible addition. I cannot think of any easy attacks since it's only the directly conflicting transactions and not their descendants (i.e. I don't have the concerns stated in https://github.com/bitcoin/bitcoin/pull/23121#issuecomment-1004774370).

Btw, don't we want min(package feerate, child ancestor feerate)?

In pseudocode for a child-with-unconfirmed-parents package: 
```
for each tx in directly_conflicting_transactions:
    if min(package_feerate, child.ancestor_feerate) <= tx.individual_feerate
        fail
```

With this, do you think it would be acceptable to remove the "no new unconfirmed inputs" rule in package RBF?


> we compare the feerate of a new transaction coming in to the feerate of the existing transactions that it directly conflicts with

Right, this wasn't in BIP 125. I also misinterpreted that part of the code to be part of the other fee-related rules and just a "fail fast" piece of logic, and didn't include it in # #23711 either (gah!). Just opened #25382 to add test coverage for it in isolation + add it to doc/policy/replacements.md

Reply:
> it's only the directly conflicting transactions and not their descendants

Could you clarify that? I'm not sure whether @sdaftuar's example contradicts this claim or not.

Let's consider a mempool containing transaction A (1000vb, 1sat/vb), child B (1000vb, 10sat/vb) and grand-child C (100vb, 100sat/vb).

If we want to replace this with a package containing transaction A' (1000vb, 1sat/vb) and child B' (1000vb, 20sat,vb), would that work? Or would we be forced to match the 100 sat/vb of grand-child C (since @sdaftuar says the ancestor feerate of that package must be higher than the _actual_ feerate of _all_ transactions that would be evicted)? If so, that would be an issue, an attacker could always put a low-fee high-feerate descendant to make it very expensive to replace the first transaction in the chain, can't they?

Reply:
> > require the ancestor feerate of the child transaction to be higher than the actual feerate of all transactions that would be evicted. This is conservative but I think would ensure that we don't have any situations where there are transactions being replaced that might appear in the next block, while the new transactions do not.
> 
> I think this is a sensible addition. I cannot think of any easy attacks since it's only the directly conflicting transactions and not their descendants (i.e. I don't have the concerns stated in [#23121 (comment)](https://github.com/bitcoin/bitcoin/pull/23121#issuecomment-1004774370)).

Ah, in writing this I overlooked what you had written up in that comment and forgot about that concern.  When I commented before I did indeed mean to suggest that we'd compare our estimated mining score of the new child transaction (`min(package feerate, ancestor feerate)` as you say) with the actual feerate of all the transactions that would be evicted, and not just the direct conflicts -- if we only compare to the feerate of direct conflicts, then I believe it is still possible for a transaction that would be included in the next block to be replaced by a transaction that will confirm much later.

However now you have reminded me of the other concern, which is that a high actual feerate (but low ancestor feerate) transaction could be used to pin a package and make it difficult to evict.  I'm not sure what to do about this!  

Perhaps a simple, good enough approach might be to require that the child transaction's score is better than the ancestor feerate of all the transactions that would be evicted, and the actual feerates of the direct conflicts?  I just think we should do something to ensure that a new transaction package isn't strictly worse than something that would be evicted.  We'd still be missing the case where something being evicted has a lower ancestor score than its actual mining score because of an unrelated transaction that is paying for a low feerate parent, but maybe that is an edge case that we can overlook (especially since the total fees in the mempool are still going up and the first order approximation of mining desirability is also going up).  

I was trying to think about whether this approach might be attackable in some way to either muck with fee estimation or attack someone's mempool with low quality transactions, but I need to give it more thought.

An alternate approach to this problem might be to define some new (opt in) relay policy that limits the scope of descendant transactions in some way, and then only implement a package RBF policy that would interact with transactions that meet that limited criteria.  That might be a good direction to go if our broader attempts at this problem result in behavior that's not incentive compatible..?

Reply:
>  When I commented before I did indeed mean to suggest that we'd compare our estimated mining score of the new child ...
> However now you have reminded me of the other concern, which is that a high actual feerate (but low ancestor feerate) transaction could be used to pin a package and make it difficult to evict. I'm not sure what to do about this!

My bad about misunderstanding your original suggestion! Yes, this is my concern with it and I'm glad you agree.

> Perhaps a simple, good enough approach might be to require that the child transaction's score is better than the ancestor feerate of all the transactions that would be evicted, and the actual feerates of the direct conflicts?

I think this would be fine, definitely an improvement upon our current logic. I can't really think of a strong attack, will think about it more.

> Let's say you have a mempool containing parent A ...
> While the ancestor score of C' is better than C, the feerate is far less compelling and it should be rejected by the mempool.

IIUC since we use the minimum of ancestor feerate and individual feerate, this example doesn't break it, since we'd use `C'`'s individual feerate of 10sat/vB to compare to `C`'s individual feerate of 20sat/vB.

The natural "attack" scenario might be:
A counterparty broadcasts their tx A (which we'll want to replace). They also broadcast an unrelated high-fee tx B (50sat/vB), and a child C which spends from A and B. Since C spends from B, it has a high ancestor feerate (even though they wouldn't be mined together), and is thus harder to replace. But since B is high feerate, it will be mined soon, so we're probably fine.

> An alternate approach to this problem might be to define some new (opt in) relay policy that limits the scope of descendant transactions in some way, and then only implement a package RBF policy that would interact with transactions that meet that limited criteria. That might be a good direction to go if our broader attempts at this problem result in behavior that's not incentive compatible..?

I'm now pretty convinced that opt-in shrinking ancestor/descendant limits is the way to go (ping @instagibbs who had ideas about this being necessary in the future). Maybe we start allowing nVersion=3 (or whatever signal we decide) and it doubles as a signal for package RBF and stricter limits (and BIP 125 signaling would not be used for package RBF).

A few questions:

- Is it possible to have single transaction limits as well (e.g. a single transaction cannot be more than 10KvB)? If so, that seems simpler than restricting descendants based on the size of the original transaction. Or would that be a bad idea, since it would also imply fewer possible in-flight HTLCs?
- In this policy, I think we would want to require "any tx spending from an unconfirmed nVersion=3 tx must also be nVersion=3?" And if a nVersion=3 transaction has ancestors that aren't nVersion=3, we'd use the stricter limits? Following that, I guess we would also need "any tx replacing an nVersion=3 tx must also be nVersion=3?"
- Assuming we can figure out something sensible, it seems better to have both ancestor and descendant limits? Adding descendant limits resolves the "you must pay for all conflicts" type of pinning. But adding an ancestor limit could be beneficial in resolving the "create a replacement tx with a giant ancestor" type pinning.

Reply:
> Assuming we can figure out something sensible, it seems better to have both ancestor and descendant limits? Adding descendant limits resolves the "you must pay for all conflicts" type of pinning. But adding an ancestor limit could be beneficial in resolving the "create a replacement tx with a giant ancestor" type pinning.

Yeah I was thinking in this direction as well.  I wonder if we could find some super minimal use case, like a 2-transaction package replacing another 2-transaction package, and just implement that to start.  I think the rough idea would be:
- some opt-in flag (eg nVersion=3, or anything else) indicates that a transaction can't have more than 1 in-mempool descendant with no more than X vbytes of total size.  Any descendant is also not allowed to have more than 1 ancestor.
- When a 1-child, 1-parent package comes in, if it conflicts with at most 1 opt-in package, allow for RBF semantics along the lines of what we're discussing. In this case, I think we would require the package feerate to be greater than the ancestor feerates of both transactions that would be replaced (along with the total fee check, and the incremental relay fee check).

Would there be value to any layer 2 projects if something this minimal were to be implemented?

Reply:
This would be very valuable for lightning. @instagibbs has explored that space recently.

A single descendant is too limiting though: in the lightning case, each participant has its own anchor output, so we need to allow 2 descendants if we apply that rule to the commitment transaction.

Reply:
> [1] In looking at this example, I noticed that BIP 125 as drafted seems to misstate an important check we perform in our current RBF policy, namely that we compare the feerate of a new transaction coming in to the feerate of the existing transactions that it directly conflicts with.

Perhaps we should aim for a new rbf bip that describes bitcoin core's actual policy for both rbf of single txs and packages of txs, and document that we implement the new bip, and not 125?